PROTOCOL.md

```diff
@@ -0,0 +1,74 @@
+# GraphQL over WebSocket Protocol
+
+## Client-server communication
+
+Each message has a `type` field, which defined in the protocol of this package, as well as associated fields inside `payload` field, depending on the message type.
+
+### Client -> Server
+
+#### GQL_CONNECTION_INIT
+Client sends this message after connecting, this triggers `onConnect` on the server.
+- `payload: Object` : optional parameters that the client specifies in `connectionParams`
+
+#### GQL_CONNECTION_TERMINATE
+Client sends this message to terminate the connection.
+
+#### GQL_START
+Client sends this message to execute GraphQL operation
+- `query: string` : GraphQL operation as string or parsed GraphQL document node
+- `variables?: Object` : Object with GraphQL variables
+- `operationName?: string` : GraphQL operation name
+- `context?: any` : Execution context for the operation
+    
+#### GQL_STOP
+Client sends this message in order to stop a running GraphQL operation execution (for example: unsubscribe)
+- `id: string` : operation id
+    
+### Server -> Client
+
+#### GQL_CONNECTION_ERROR
+The server sends this message if `onConnect` callback returns `false` or throws an exception. after sending this message, the server disconnects the client.
+- `payload: Object`: the server side error
+
+#### GQL_CONNECTION_ACK
+The server sends this message if `onConnect` callback returns any other value then `false`.
+
+#### GQL_DATA
+The server sends this message to transfter the GraphQL execution result from the server to the client.
+- `id: string` : ID of the operation that was successfully set up
+- `payload: { data: any }` : Execution result
+- `payload: { errors?: [] }` : Array of resolvers errors
+
+#### GQL_ERROR
+Server sends this message upon failing to **register** or **execute** a GraphQL operation (resolver errors are part of `GQL_DATA` message, and will be added as `errors` array)
+- `payload: { errors: Array<Object> }` : payload with an array of errors attributed to the subscription failing on the server
+- `id: string` : operation ID of the operation that failed on the server
+
+#### GQL_COMPLETE
+Server sends this message to indicate that a GraphQL operation is done, and no more data will arrive (for queries and subscription, it called immediatly, and for subscriptions it called when unsubscribing)
+- `id: string` : operation ID of the operation that completed
+
+#### GQL_CONNECTION_KEEP_ALIVE
+Server message sent periodically to keep the client connection alive.
```

Comment:
It would be interesting mention here that the client starts to considerer the keep live only upon the first received keep alive message from the server.